p2p/shim.go

```diff
@@ -0,0 +1,327 @@
+// The embedding reactor must be sure to listen for messages on this channel to
+// handle adding a peer.
+func (rs *ReactorShim) AddPeer(peer Peer) {
+	peerID, err := PeerIDFromString(string(peer.ID()))
+	if err != nil {
+		rs.Logger.Error("failed to add peer", "peer", peer.ID(), "err", err)
+		return
+	}
+
+	rs.PeerUpdateCh <- PeerUpdate{PeerID: peerID, Status: PeerStatusUp}
+	rs.Logger.Debug("sent peer update", "reactor", rs.Name, "peer", peerID.String(), "status", PeerStatusUp)
+}
+
+// RemovePeer sends a PeerUpdate with status PeerStatusDown on the PeerUpdateCh.
+// The embedding reactor must be sure to listen for messages on this channel to
+// handle removing a peer.
+func (rs *ReactorShim) RemovePeer(peer Peer, reason interface{}) {
+	peerID, err := PeerIDFromString(string(peer.ID()))
+	if err != nil {
+		rs.Logger.Error("failed to remove peer", "peer", peer.ID(), "err", err)
+		return
+	}
+
+	rs.PeerUpdateCh <- PeerUpdate{PeerID: peerID, Status: PeerStatusDown}
+	rs.Logger.Debug(
+		"sent peer update",
+		"reactor", rs.Name,
+		"peer", peerID.String(),
+		"reason", reason,
+		"status", PeerStatusDown,
+	)
+}
+
+// Receive implements a generic wrapper around implementing the Receive method
+// on the legacy Reactor p2p interface. If the reactor is running, Receive will
+// find the corresponding new p2p Channel, create and decode the appropriate
+// proto.Message from the msgBytes, execute any validation and finally construct
+// and send a p2p Envelope on the appropriate p2p Channel.
+func (rs *ReactorShim) Receive(chID byte, src Peer, msgBytes []byte) {
+	if !rs.IsRunning() {
+		return
+	}
+
+	cID := ChannelID(chID)
+	channelShim, ok := rs.Channels[cID]
+	if !ok {
+		rs.Logger.Error("unexpected channel", "peer", src, "ch_id", chID)
+		return
+	}
+
+	peerID, err := PeerIDFromString(string(src.ID()))
+	if err != nil {
+		rs.Logger.Error("failed to convert peer ID", "peer", src, "ch_id", chID, "err", err)
+		return
+	}
+
+	msg := proto.Clone(channelShim.Channel.messageType)
+	msg.Reset()
+
+	if err := proto.Unmarshal(msgBytes, msg); err != nil {
+		rs.Logger.Error("error decoding message", "peer", src, "ch_id", cID, "msg", msg, "err", err)
+		rs.Switch.StopPeerForError(src, err)
+		return
+	}
+
+	validator, ok := msg.(messageValidator)
+	if ok {
+		if err := validator.Validate(); err != nil {
+			rs.Logger.Error("invalid message", "peer", src, "ch_id", cID, "msg", msg, "err", err)
+			rs.Switch.StopPeerForError(src, err)
+			return
+		}
+	}
+
+	wrapper, ok := msg.(Wrapper)
+	if ok {
+		var err error
+
+		msg, err = wrapper.Unwrap()
+		if err != nil {
+			rs.Logger.Error("failed to unwrap message", "peer", src, "ch_id", chID, "msg", msg, "err", err)
+			return
+		}
+	}
+
+	select {
+	case channelShim.InCh <- Envelope{From: peerID, Message: msg}:
+		rs.Logger.Debug("proxied envelope", "reactor", rs.Name, "ch_id", cID, "peer", peerID.String())
+
+	case <-channelShim.Channel.Done():
+		close(channelShim.InCh)
```

Comment:
I don't believe it's safe to close this channel here, since other goroutines may currently be blocked on sending to it and the Go scheduler may try resuming the send before trying to read the Done channel.

I suppose we could simply say that the `In` channel will never be closed, but rather that the senders will eventually give up (once they check `Done`), making it the consumer's responsibility to stop listening on `In` once `Done` returns. Otherwise we will need additional synchronization here to make sure we only close it after guaranteeing that no other goroutines will write to it. I feel like closing it is makes for a better API with fewer footguns, but I'm not sure how tricky it will be to get it right.

Reply:
Ughhh yeah, you're right -- we can't close here, especially if `In` is not buffered (which it's not currently).

> Otherwise we will need additional synchronization here to make sure we only close it after guaranteeing that no other goroutines will write to it. 

I cannot immediately think of any way of doing this apart from using a mutex that wraps the `select` so only one goroutine has write access at a time. But isn't that an indication of a poor design choice -- locks with channel writes/closing?

Reply:
>I cannot immediately think of any way of doing this apart from using a mutex that wraps the select so only one goroutine has write access at a time. But isn't that an indication of a poor design choice -- locks with channel writes/closing?

It's not quite that bad -- we could for example use a waitgroup, but that probably wouldn't always work since it might never go down to 0 thus the channel could end up never being closed. This sort of synchronization also has an additional performance cost. There are ways to solve this, but they're not trivial and would need some careful thought.

This is basically the M-to-N scenario outlined in the link we discussed last week.

Reply:
Why don't we just use a lock and replace `In` with a no-op consumer? This way we don't have goroutines piling up and we don't have to worry about synchronization outside of just using a lock.

Reply:
Decided based on our convo to explicitly NOT close the inbound channel. I've documented this in numerous places. Resolving for now. We can revisit this at a later time if we want.